content/blogposts/2023/building-conc-better-structured-concurrency-for-go.md

```diff
@@ -0,0 +1,326 @@
+* [Making concurrent code easier to read](https://github.com/sourcegraph/conc/blob/main/README.md#goal-3-make-concurrent-code-easier-to-read)
+
+
+## Using conc for parallel stream processing
+
+Something else conc does taht is more specific to our usecase as Sourcegraph is parallel stream processing. While indexing large amounts of code, we often end up with multiple streams of results that we want to post-process. Each result in the stream requires at least one network request, for example to look up permission on a repository or to fetch the rest of a file for a specific query match.
+
+For this we always want to:
+
+* Do the network requests in parallel
+* Show the results to the user as soon as possible
+* Maintain the order of the stream (beacuse we've already ranked the results)
+
+It's difficult to get all three of these right at the same time, so one of the goals I had while writing [Conc's Stream package](https://pkg.go.dev/github.com/sourcegraph/conc/stream#Stream) is to abstract away as much of the complexity for that workflow as possible.
+
+Now I can write custom functions to handle multiple file streams at once using code simliar to the example below. This efficiently and safely gets the contents of each file from a list of file names.
+
+```go
+func streamFileContents(ctx context.Context, fileNames <-chan string, fileContents chan<- string) {
+	s := stream.New()
+	for fileName := range fileNames {
+		fileName := fileName
+		s.Go(func() stream.Callback {
+			contents := fetchFileContents(ctx, fileName)
+			return func() { fileContents <- contents }
+		})
+	}
+	s.Wait()
+}
+```
+
+## The goals of `conc`
+
+Conc is hopefully already useful to anyone writing concurrent Go code who wants a better way to handle panics, avoid leaking Goroutines, or simply have more readable concurrent code. It's still a young project though and I expect it to evolve and improve over time. One thing I've had to think about is what I *don't* want it to be. 
+
+My main goal with `conc` is that it should be hard to use incorrectly. Concurrent code is infamous for causing headaches and hard-to-identify bugs, and while many developers might feel like they are smart enough to avoid the problems associated with concurrent programming, in reality these bugs can burn anyone.
+
+![Puppies parallel programming](https://storage.googleapis.com/sourcegraph-assets/blog/upload_ee1c4226c42ba701cde2324025599202.png)
+
+This goal of "make it hard to use incorrectly" comes with tradeoffs. Specifically, it means some (potentially useful) features are unlikely to be added to conc because they would break this goal. Shortly after the library went public I got a request to [add channels to conc](https://github.com/sourcegraph/conc/issues/56) and while I can see how some advanced user of conc could get value from this, I'd rather keep to the "hard to misuse" goal wherever possible.
```

Comment:
```suggestion
This goal of "make it hard to use incorrectly" comes with tradeoffs. Specifically, it means some (potentially useful) features are unlikely to be added to conc because they would break this goal. Shortly after the library went public I got a request to [add channels to conc](https://github.com/sourcegraph/conc/issues/56) and while I can see how a user of conc could find this valuable, I'd like to keep things "hard to misuse" wherever possible.
```